server/schema/strongly_typed_marshaller.go

```
@@ -0,0 +1,102 @@
+	"github.com/xitongsys/parquet-go/types"
+	"github.com/xitongsys/parquet-go/writer"
+	"time"
+)
+
+type StronglyTypedMarshaller interface {
+	Marshal(bh *BatchHeader, data []map[string]interface{}) ([]byte, error)
+}
+
+func NewParquetMarshaller() StronglyTypedMarshaller {
+	pm := &ParquetMarshaller{
+		GoroutinesCount: 2,
+	}
+	return pm
+}
+
+type ParquetMarshaller struct {
+	GoroutinesCount int64
+}
+
+func (pm *ParquetMarshaller) Marshal(bh *BatchHeader, data []map[string]interface{}) ([]byte, error) {
+	md, fieldIndex := pm.parquetMetadata(bh)
+	buf := new(bytes.Buffer)
+	fw := writerfile.NewWriterFile(buf)
+	defer fw.Close()
+	pw, err := writer.NewCSVWriter(md, fw, pm.GoroutinesCount)
+	if err != nil {
+		return nil, fmt.Errorf("can't create parquet writer: %v", err)
+	}
+	defer func() {
+		if err = pw.WriteStop(); err != nil {
+			logging.Errorf("parquet writeStop error: %v", err)
+		}
+	}()
+	for _, obj := range data {
+		parquetRec := pm.parquetRecord(bh, obj, fieldIndex)
+		if err = pw.Write(parquetRec); err != nil {
+			logging.Warnf("parquet write error: %v", err)
+		}
+	}
+	return buf.Bytes(), nil
+}
+
+func (pm *ParquetMarshaller) parquetMetadata(bh *BatchHeader) ([]string, map[string]int) {
+	md := make([]string, 0, len(bh.Fields))
+	fieldIndex := make(map[string]int, len(bh.Fields))
+	i := 0
+	for field, fieldMeta := range bh.Fields {
+		fieldIndex[field] = i
+		switch *fieldMeta.dataType {
+		case typing.BOOL:
+			md = append(md, fmt.Sprintf("name=%s, type=BOOLEAN", field))
+		case typing.INT64:
+			md = append(md, fmt.Sprintf("name=%s, type=INT64", field))
+		case typing.FLOAT64:
+			md = append(md, fmt.Sprintf("name=%s, type=DOUBLE", field))
+		case typing.STRING:
+			md = append(md, fmt.Sprintf("name=%s, type=BYTE_ARRAY, convertedtype=UTF8, encoding=PLAIN_DICTIONARY", field))
+		case typing.TIMESTAMP:
+			md = append(md, fmt.Sprintf("name=%s, type=TIMESTAMP_MILLIS", field))
+		// UNKNOWN and default
```

let's return an error if unknown type

---

done